hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/associations/TokenAssociationsDumpUtils.java

```diff
@@ -0,0 +1,150 @@
+
+    @NonNull
+    private static <K extends VirtualKey, V extends VirtualValue>
+            Map<TokenAssociationId, TokenAssociation> gatherTokenRelations(
+                    @NonNull final VirtualMap<K, V> source,
+                    @NonNull final Function<K, TokenAssociationId> keyMapper,
+                    @NonNull final Function<V, TokenAssociation> valueMapper) {
+        final var r = new HashMap<TokenAssociationId, TokenAssociation>();
+        final var threadCount = 8;
+        final var tokenAssociations = new ConcurrentLinkedQueue<Pair<TokenAssociationId, TokenAssociation>>();
+        try {
+            VirtualMapLike.from(source)
+                    .extractVirtualMapData(
+                            getStaticThreadManager(),
+                            p -> tokenAssociations.add(
+                                    Pair.of(keyMapper.apply(p.left()), valueMapper.apply(p.right()))),
+                            threadCount);
+        } catch (final InterruptedException ex) {
+            System.err.println("*** Traversal of token associations virtual map interrupted!");
+            Thread.currentThread().interrupt();
+        }
+        while (!tokenAssociations.isEmpty()) {
+            final var mapping = tokenAssociations.poll();
+            r.put(mapping.key(), mapping.value());
+        }
+        return r;
+    }
+
+    private static void reportOnUniques(
+            @NonNull final Writer writer, @NonNull final Map<TokenAssociationId, TokenAssociation> associations) {
+        writer.writeln(formatHeader());
+        associations.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .forEach(e -> formatUnique(writer, e.getValue()));
+        writer.writeln("");
+    }
+
+    @NonNull
+    private static String formatHeader() {
+        return "nftId,nftSerial,"
+                + fieldFormatters.stream().map(Pair::left).collect(Collectors.joining(Writer.FIELD_SEPARATOR));
+    }
+
+    // spotless:off
+    @NonNull
+    private static final List<Pair<String, BiConsumer<FieldBuilder, TokenAssociation>>> fieldFormatters = List.of(
+            Pair.of("owner", getFieldFormatter(TokenAssociation::accountId, ThingsToStrings::toStringOfEntityId)),
+            Pair.of("spender", getFieldFormatter(TokenAssociation::tokenId, ThingsToStrings::toStringOfEntityId)),
+            Pair.of("creationTime", getFieldFormatter(TokenAssociation::balance, Object::toString)),
+            Pair.of("metadata", getFieldFormatter(TokenAssociation::isFrozen, Object::toString)),
+            Pair.of("prev", getFieldFormatter(TokenAssociation::isKycGranted, Object::toString)),
+            Pair.of("next", getFieldFormatter(TokenAssociation::isAutomaticAssociation, Object::toString)),
+            Pair.of("next", getFieldFormatter(TokenAssociation::prev, ThingsToStrings::toStringOfEntityId)),
+            Pair.of("next", getFieldFormatter(TokenAssociation::next, ThingsToStrings::toStringOfEntityId))
```

Comment:
The field names "owner", "spender", etc are all wrong for associations.  This means the CSV header line will not be correctd.